### PR TITLE
Us 709 voltage level dialog

### PR DESCRIPTION
**Pull request by laurgarn**

### CODE REVIEW

File: src/main/java/org/gridsuite/modification/server/entities/equipment/creation/VoltageLevelCreationEntity.java [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+
+import org.gridsuite.modification.server.ModificationType;
+import org.gridsuite.modification.server.dto.BusbarConnectionCreationInfos;
+import org.gridsuite.modification.server.dto.BusbarSectionCreationInfos;
+import org.gridsuite.modification.server.dto.EquipmenModificationInfos;
+import org.gridsuite.modification.server.dto.VoltageLevelCreationInfos;
+
+import lombok.Getter;
+import lombok.NoArgsConstructor;
+
```
Comment (FranckLecuyer) [inline src/main/java/org/gridsuite/modification/server/entities/equipment/creation/VoltageLevelCreationEntity.java:29]: Missing @author

File: src/main/java/org/gridsuite/modification/server/service/NetworkModificationService.java [hunk 2]
```diff
@@ -1028,6 +1029,134 @@ private TwoWindingsTransformer createTwoWindingsTransformer(Network network, Vol
...
     }
 
+    private List<EquipmenModificationInfos> execCreateVoltageLevel(
+        NetworkStoreListener listener, VoltageLevelCreationInfos voltageLevelCreationInfos,
+        ReporterModel reporter,
+        Reporter subReporter) {
+
+        Network network = listener.getNetwork();
+        UUID networkUuid = listener.getNetworkUuid();
+        Substation substation = network.getSubstation(voltageLevelCreationInfos.getSubstationId());
```
Comment (FranckLecuyer) [inline src/main/java/org/gridsuite/modification/server/service/NetworkModificationService.java:1039]: Should add a test if substation exists

File: src/main/java/org/gridsuite/modification/server/service/NetworkModificationService.java [hunk 3]
```diff
@@ -1028,6 +1029,134 @@ private TwoWindingsTransformer createTwoWindingsTransformer(Network network, Vol
...
+                    throw new NetworkModificationException(CREATE_VOLTAGE_LEVEL_ERROR, "To side '" + toBBSId + "' unknown");
+                }
+
+                SwitchKind switchKind = bbsci.getSwitchKind();
+                if (switchKind == SwitchKind.BREAKER) {
+                    int preBreakerRank = nodeRank++;
+                    int postBreakerRank = nodeRank++;
+                    voltageLevel.getNodeBreakerView().newDisconnector()
+                        .setKind(SwitchKind.DISCONNECTOR)
+                        .setId("switch_" + voltageLevelCreationInfos.getEquipmentId() + "_" + fromBBSId + "_" + toBBSId
```
Comment (FranckLecuyer) [inline src/main/java/org/gridsuite/modification/server/service/NetworkModificationService.java:1085]: maybe "disconnector_" , like in createNodeBreakerCellSwitches method ?

File: src/main/java/org/gridsuite/modification/server/service/NetworkModificationService.java [hunk 4]
```diff
@@ -1028,6 +1029,134 @@ private TwoWindingsTransformer createTwoWindingsTransformer(Network network, Vol
...
+                        .setNode1(preBreakerRank)
+                        .setNode2(postBreakerRank)
+                        .setFictitious(false)
+                        .setRetained(false)
+                        .setOpen(false)
+                        .add();
+
+                    voltageLevel.getNodeBreakerView().newDisconnector()
+                        .setKind(SwitchKind.DISCONNECTOR)
+                        .setId("switch_" + voltageLevelCreationInfos.getEquipmentId() + "_" + toBBSId + "_" + fromBBSId
```
Comment (FranckLecuyer) [inline src/main/java/org/gridsuite/modification/server/service/NetworkModificationService.java:1106]: maybe "disconnector_" , like in createNodeBreakerCellSwitches method ?

File: src/main/java/org/gridsuite/modification/server/service/NetworkModificationService.java [hunk 5]
```diff
@@ -1028,6 +1029,134 @@ private TwoWindingsTransformer createTwoWindingsTransformer(Network network, Vol
...
+                        .setNode1(rank1)
+                        .setNode2(preBreakerRank)
+                        .setFictitious(false)
+                        .setRetained(false)
+                        .setOpen(false)
+                        .add();
+
+                    voltageLevel.getNodeBreakerView().newBreaker()
+                        .setKind(switchKind)
+                        .setId("breaker_" + voltageLevelCreationInfos.getEquipmentId() + "_" + cnxRank++)
```
Comment (FranckLecuyer) [inline src/main/java/org/gridsuite/modification/server/service/NetworkModificationService.java:1096]: maybe "breaker_" , like in createNodeBreakerCellSwitches method ?

And you can fromBBSId and toBBSId to construct the id, like for the 2 disconnectors

File: src/main/java/org/gridsuite/modification/server/service/NetworkModificationService.java [hunk 6]
```diff
@@ -1028,6 +1029,134 @@ private TwoWindingsTransformer createTwoWindingsTransformer(Network network, Vol
...
+                        .setNode1(postBreakerRank)
+                        .setNode2(rank2)
+                        .setFictitious(false)
+                        .setRetained(false)
+                        .setOpen(false)
+                        .add();
+                } else {
+                    voltageLevel.getNodeBreakerView().newDisconnector()
+                        .setKind(switchKind)
+                        .setId("switch_" + voltageLevelCreationInfos.getEquipmentId() + "_" + fromBBSId + "_" + toBBSId
```
Comment (FranckLecuyer) [inline src/main/java/org/gridsuite/modification/server/service/NetworkModificationService.java:1117]: maybe "disconnector_" , like in createNodeBreakerCellSwitches method ?

File: src/main/java/org/gridsuite/modification/server/service/NetworkModificationService.java [hunk 7]
```diff
@@ -1028,6 +1029,134 @@ private TwoWindingsTransformer createTwoWindingsTransformer(Network network, Vol
...
+                            + "_" + cnxRank++)
+                        .setNode1(rank1)
+                        .setNode2(rank2)
+                        .setFictitious(false)
+                        .setRetained(false)
+                        .setOpen(false)
+                        .add();
+                }
```
Comment (FranckLecuyer) [inline src/main/java/org/gridsuite/modification/server/service/NetworkModificationService.java:1125]: Maybe throw an exception if SwitchKind is not BREAKER or DISCONNECTOR ?

File: src/main/java/org/gridsuite/modification/server/service/NetworkModificationService.java [hunk 8]
```diff
@@ -1028,6 +1029,134 @@ private TwoWindingsTransformer createTwoWindingsTransformer(Network network, Vol
...
+                        .setKind(SwitchKind.DISCONNECTOR)
+                        .setId("switch_" + voltageLevelCreationInfos.getEquipmentId() + "_" + toBBSId + "_" + fromBBSId
+                            + "_" + cnxRank++)
+                        .setNode1(postBreakerRank)
+                        .setNode2(rank2)
+                        .setFictitious(false)
+                        .setRetained(false)
+                        .setOpen(false)
+                        .add();
+                } else {
```
Comment (FranckLecuyer) [inline src/main/java/org/gridsuite/modification/server/service/NetworkModificationService.java:1114]: else if (switchKind == SwitchKind.DISCONNECTOR) ?

File: src/test/java/org/gridsuite/modification/server/utils/MatcherVoltageLevelCreationInfos.java [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+    }
+
+    public static MatcherVoltageLevelCreationInfos createMatcherVoltageLevelCreationInfos(VoltageLevelCreationInfos ref) {
+        return new MatcherVoltageLevelCreationInfos(ref);
+    }
+
+    public boolean matchesSafely(VoltageLevelCreationInfos m) {
+        return super.matchesSafely(m)
+            && Math.abs(reference.getNominalVoltage() - m.getNominalVoltage()) < 0.2
+            && Objects.equals(reference.getSubstationId(), m.getSubstationId());
```
Comment (FranckLecuyer) [inline src/test/java/org/gridsuite/modification/server/utils/MatcherVoltageLevelCreationInfos.java:25]: No test on busbarSections and busbarConnections ?

**Review by FranckLecuyer — CHANGES_REQUESTED** [review]

File: src/main/java/org/gridsuite/modification/server/dto/BusbarConnectionCreationInfos.java [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+import lombok.Setter;
+import lombok.experimental.SuperBuilder;
+
+@SuperBuilder
+@NoArgsConstructor
+@AllArgsConstructor
+@Getter
+@Setter
+@Schema(description = "Voltage level bus bar sections connection creation")
+public class BusbarConnectionCreationInfos {
```
[the remark this answers is not in the document]
Comment (laurgarn) [reply src/main/java/org/gridsuite/modification/server/dto/BusbarConnectionCreationInfos.java:23]: 29 chars by now, then would be 36. It starts to get quite long, doesn't it ?

File: src/main/java/org/gridsuite/modification/server/dto/BusbarConnectionCreationInfos.java [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+package org.gridsuite.modification.server.dto;
+
+import com.powsybl.iidm.network.SwitchKind;
+import io.swagger.v3.oas.annotations.media.Schema;
+import lombok.AllArgsConstructor;
+import lombok.Getter;
+import lombok.NoArgsConstructor;
+import lombok.Setter;
+import lombok.experimental.SuperBuilder;
+
```
[the remark this answers is not in the document]
Comment (laurgarn) [reply src/main/java/org/gridsuite/modification/server/dto/BusbarConnectionCreationInfos.java:16]: done

File: src/main/java/org/gridsuite/modification/server/dto/VoltageLevelCreationInfos.java [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+
+import java.util.List;
+
+import io.swagger.v3.oas.annotations.media.Schema;
+import lombok.Getter;
+import lombok.NoArgsConstructor;
+import lombok.Setter;
+import lombok.ToString;
+import lombok.experimental.SuperBuilder;
+
```
[the remark this answers is not in the document]
Comment (laurgarn) [reply src/main/java/org/gridsuite/modification/server/dto/VoltageLevelCreationInfos.java:17]: done

File: src/main/java/org/gridsuite/modification/server/entities/equipment/creation/BusbarConnectionCreationEmbeddable.java [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+ */
+package org.gridsuite.modification.server.entities.equipment.creation;
+
+import javax.persistence.Embeddable;
+
+import com.powsybl.iidm.network.SwitchKind;
+import lombok.AllArgsConstructor;
+import lombok.Getter;
+import lombok.NoArgsConstructor;
+
```
[the remark this answers is not in the document]
Comment (laurgarn) [reply src/main/java/org/gridsuite/modification/server/entities/equipment/creation/BusbarConnectionCreationEmbeddable.java:15]: done

File: src/main/java/org/gridsuite/modification/server/entities/equipment/creation/BusbarSectionCreationEmbeddable.java [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+package org.gridsuite.modification.server.entities.equipment.creation;
+
+import javax.persistence.Embeddable;
+
+import lombok.AllArgsConstructor;
+import lombok.Getter;
+import lombok.NoArgsConstructor;
+
```
[the remark this answers is not in the document]
Comment (laurgarn) [reply src/main/java/org/gridsuite/modification/server/entities/equipment/creation/BusbarSectionCreationEmbeddable.java:14]: done

File: src/main/java/org/gridsuite/modification/server/entities/equipment/creation/VoltageLevelCreationEntity.java (hunk 1 above)
Comment (laurgarn) [reply src/main/java/org/gridsuite/modification/server/entities/equipment/creation/VoltageLevelCreationEntity.java:29]: done

File: src/main/java/org/gridsuite/modification/server/service/NetworkModificationService.java (hunk 2 above)
Comment (laurgarn) [reply src/main/java/org/gridsuite/modification/server/service/NetworkModificationService.java]: done

File: src/main/java/org/gridsuite/modification/server/service/NetworkModificationService.java (hunk 3 above)
Comment (laurgarn) [reply src/main/java/org/gridsuite/modification/server/service/NetworkModificationService.java]: done

File: src/main/java/org/gridsuite/modification/server/service/NetworkModificationService.java (hunk 4 above)
Comment (laurgarn) [reply src/main/java/org/gridsuite/modification/server/service/NetworkModificationService.java]: done

File: src/main/java/org/gridsuite/modification/server/service/NetworkModificationService.java (hunk 5 above)
Comment (laurgarn) [reply src/main/java/org/gridsuite/modification/server/service/NetworkModificationService.java]: done

File: src/main/java/org/gridsuite/modification/server/service/NetworkModificationService.java (hunk 6 above)
Comment (laurgarn) [reply src/main/java/org/gridsuite/modification/server/service/NetworkModificationService.java]: done

File: src/main/java/org/gridsuite/modification/server/service/NetworkModificationService.java (hunk 7 above)
Comment (laurgarn) [reply src/main/java/org/gridsuite/modification/server/service/NetworkModificationService.java:1130]: done

File: src/main/java/org/gridsuite/modification/server/service/NetworkModificationService.java (hunk 8 above)
Comment (laurgarn) [reply src/main/java/org/gridsuite/modification/server/service/NetworkModificationService.java]: done

File: src/test/java/org/gridsuite/modification/server/utils/MatcherVoltageLevelCreationInfos.java (hunk 9 above)
Comment (laurgarn) [reply src/test/java/org/gridsuite/modification/server/utils/MatcherVoltageLevelCreationInfos.java]: added

**Review by FranckLecuyer — APPROVED** [review]

**Pull request merged**